src/accelerate/data_loader.py

```diff
@@ -453,7 +463,11 @@ def _fetch_batches(self, iterator):
 
     def __iter__(self):
         self.gradient_state._set_end_of_dataloader(False)
-        self.gradient_state._set_samples_seen(0)
+        try:
+            length = getattr(self.dataset, "total_dataset_length", len(self.dataset))
+            self.gradient_state._set_remainder(length % self.batch_size)
```

Comment:
```suggestion
self.gradient_state._set_remainder(length % self.total_batch_size)
```

Reply:
After the above change or without it, `test_metrics.py` fails when `dispatch_batches=True` when the `num_samples % num_processes != 0` , e.g., num_samples=97 and num_processes=2. Below is the modified test with print statements. I think, `IterableDatasetShard` should be used in conjunction with  `dispatch_batches=True` for expected behaviour or `DataLoaderDispatcher` needs to be fixed to handle normal dataloaders. 

![Screenshot 2022-07-28 at 10 54 03 AM](https://user-images.githubusercontent.com/13534540/181427059-38d54e57-894e-4ae9-bf6c-9add9549c1a5.png)

Reply:
Thanks, will change the test case slightly for this and resolve today!